usdai/main.py

```diff
@@ -0,0 +1,135 @@
+import os
+import sys
+import requests
+import json
+from web3 import Web3
+from dotenv import load_dotenv
+
+from utils.chains import Chain
+from utils.web3_wrapper import ChainManager
+from utils.telegram import send_telegram_message
+from utils.cache import (
+    get_last_value_for_key_from_file, 
+    write_last_value_to_file, 
+    cache_filename
+)
+
+load_dotenv()
+
+# Constants
+PROTOCOL = "usdai"
+VAULT_ADDR = Web3.to_checksum_address("0x0A1a1A107E45b7Ced86833863f482BC5f4ed82EF")
+WM_TOKEN = Web3.to_checksum_address("0x437cc33344a0b27a429f795ff6b469c72698b291")
+SUSDAI_ADDR = Web3.to_checksum_address("0x0B2b2B2076d95dda7817e785989fE353fe955ef9")
+GRAPHQL_URL = "https://protocol-api.m0.org/graphql"
+
+def main():
+    client = ChainManager.get_client(Chain.ARBITRUM)
+    
+    # Common ABI
+    erc20_abi = [
+        {"inputs": [], "name": "totalSupply", "outputs": [{"type": "uint256"}], "stateMutability": "view", "type": "function"},
+        {"inputs": [], "name": "decimals", "outputs": [{"type": "uint8"}], "stateMutability": "view", "type": "function"},
+        {"inputs": [{"type": "address"}], "name": "balanceOf", "outputs": [{"type": "uint256"}], "stateMutability": "view", "type": "function"},
+        {"inputs": [], "name": "symbol", "outputs": [{"type": "string"}], "stateMutability": "view", "type": "function"}
+    ]
+    
+    wm = client.get_contract(WM_TOKEN, erc20_abi)
+    
+    try:
+        # --- On-Chain Supply ---
+        # USDai Supply (wM held by Vault)
+        vault_shares = wm.functions.balanceOf(VAULT_ADDR).call()
+        wm_decimals = wm.functions.decimals().call()
+        usdai_supply_fmt = vault_shares / (10 ** wm_decimals)
+        
+        # 2. Get Mint Ratio via API
+        query = """
+        query {
+          mintRatio: protocolConfigs(
+            where: {key: "mint_ratio"}
+            orderBy: blockTimestamp
+            orderDirection: desc
+            first: 1
+          ) {
+            value
+          }
+        }
+        """
+        
+        mint_ratio = 10000 # Default to 1:1 (scaled 1e4 for bps) if not found
+        
+        try:
+            res = requests.post(GRAPHQL_URL, json={"query": query}, timeout=10)
+            if res.status_code == 200:
+                data = res.json().get("data", {})
+                
+                # --- Mint Ratio ---
+                configs = data.get("mintRatio", [])
+                if configs:
+                    # Mint ratio is scaled by 1e4 (e.g. 9950 = 99.5%)
+                    mint_ratio_raw = int(configs[0].get("value", 10000))
+                    mint_ratio = mint_ratio_raw
+                
```

Comment:
Why do we need to use the mint ratio from the API? How is it related to on-chain value in the vault contract?